kotlin/src/test/kotlin/io/logto/sdk/core/util/UriUtilsTest.kt

```diff
@@ -0,0 +1,135 @@
+package io.logto.sdk.core.util
+
+import com.google.common.truth.Truth.assertThat
+import io.logto.sdk.core.exception.LogtoException
+import org.junit.Assert
+import org.junit.Test
+
+class UriUtilsTest {
+    @Test
+    fun verifyAndParseCodeFromCallbackUri() {
+        val state = GenerateUtils.generateState()
+        val code = "testCode"
+        val redirectUri = "https://myapp.com/callback"
+        val callbackUri = "https://myapp.com/callback?state=$state&code=$code"
+
+        val resultCode = UriUtils.verifyAndParseCodeFromCallbackUri(callbackUri, redirectUri, state)
+
+        assertThat(resultCode).isEqualTo(code)
+    }
+
+    @Test
+    fun verifyAndParseCodeFromCallbackUriShouldThrowWithMismatchedUri() {
+        val state = GenerateUtils.generateState()
+        val code = "testCode"
+        val redirectUri = "https://myapp.com/callback"
+        val callbackUri = "https://yourapp.com/callback?state=$state&code=$code"
+
+        val expectedException = Assert.assertThrows(LogtoException.VerificationException::class.java) {
+            UriUtils.verifyAndParseCodeFromCallbackUri(callbackUri, redirectUri, state)
+        }
+
+        assertThat(expectedException).hasMessageThat().contains(LogtoException.Verification.URI_MISMATCHED.name)
+    }
+
+    @Test
+    fun verifyAndParseCodeFromCallbackUriShouldThrowWithEmptyUri() {
+        val expectedException = Assert.assertThrows(LogtoException.VerificationException::class.java) {
+            UriUtils.verifyAndParseCodeFromCallbackUri("", "", "dummyState")
+        }
+
+        assertThat(expectedException)
+            .hasMessageThat()
+            .contains(LogtoException.Verification.INVALID_URI_FORMAT.name)
+    }
+
+    @Test
+    fun verifyAndParseCodeFromCallbackUriShouldThrowWithInvalidUriFormat() {
+        val expectedException = Assert.assertThrows(LogtoException.VerificationException::class.java) {
+            UriUtils.verifyAndParseCodeFromCallbackUri("invalidUri", "invalidUri", "dummyState")
+        }
+
+        assertThat(expectedException)
+            .hasMessageThat()
+            .contains(LogtoException.Verification.INVALID_URI_FORMAT.name)
+    }
+
+    @Test
+    fun verifyAndParseCodeFromCallbackUriShouldThrowWithErrorParameter() {
+        val state = GenerateUtils.generateState()
+        val code = "dummyCode"
+        val error = "error"
+        val redirectUri = "https://myapp.com/callback"
+        val callbackUri = "https://myapp.com/callback?error=$error&state=$state&code=$code"
+
+        val expectedException = Assert.assertThrows(LogtoException.RedirectUriReturnedException::class.java) {
+            UriUtils.verifyAndParseCodeFromCallbackUri(callbackUri, redirectUri, state)
+        }
+
+        assertThat(expectedException.error).isEqualTo(error)
+        assertThat(expectedException.errorDesc).isNull()
+    }
+
+    @Test
+    fun verifyAndParseCodeFromCallbackUriShouldThrowWithBothErrorDescAndErrorParameter() {
```

Comment:
maybe rename to should show errorDesc message if present in the callback URL. As errorDesc itself takes no effect on throwing an exception right. i.e. if no error is present, just an errorDesc there won't be any exceptions at all. This case is more like to test the exception should contains errorDesc content.

Reply:
updated